File: gaiag/baseline/Comp.java
// Gaiag --- Guile in Asd In Asd in Guile.
//
// This file is part of Gaiag.
//
// Copyright © 2014, 2015 Jan Nieuwenhuizen <janneke@gnu.org>
//
// Gaiag is free software: you can redistribute it and/or modify it
// under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// Gaiag is distributed in the hope that it will be useful, but
// WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public
// License along with Gaiag.  If not, see <http://www.gnu.org/licenses/>.
//
// Commentary:
//
// Code:

class Comp extends Component {
  enum State {
    Uninitialized, Initialized, Error
  };

  State s;
  IComp.result_t reply_IComp_result_t;

  IDevice.result_t reply_IDevice_result_t;


  IComp client;
  IDevice device_A;

  public Comp(Runtime runtime) {this(runtime, "");};

  public Comp(Runtime runtime, String name) {this(runtime, name, null);};

  public Comp(Runtime runtime, String name, SystemComponent parent) {
    super(runtime, name, parent);
    this.flushes = true;
    s = State.Uninitialized;
    client = new IComp();
    client.in.name = "client";
    client.in.self = this;
    s = State.Uninitialized;
    device_A = new IDevice();
    device_A.out.name = "device_A";
    device_A.out.self = this;
    client.in.initialize = new ValuedAction<IComp.result_t>() {public IComp.result_t action() {return Runtime.callIn(Comp.this, new ValuedAction<IComp.result_t>() {public IComp.result_t action() {return client_initialize();}}, new Meta(Comp.this.client, "initialize"));};};

    client.in.recover = new ValuedAction<IComp.result_t>() {public IComp.result_t action() {return Runtime.callIn(Comp.this, new ValuedAction<IComp.result_t>() {public IComp.result_t action() {return client_recover();}}, new Meta(Comp.this.client, "recover"));};};

    client.in.perform_actions = new ValuedAction<IComp.result_t>() {public IComp.result_t action() {return Runtime.callIn(Comp.this, new ValuedAction<IComp.result_t>() {public IComp.result_t action() {return client_perform_actions();}}, new Meta(Comp.this.client, "perform_actions"));};};

  };
  public IComp.result_t client_initialize() {
    if (s == State.Uninitialized) {
      {
        V<IDevice.result_t> res = new V <IDevice.result_t>(device_A.in.initialize.action());
        if (res.v == IDevice.result_t.OK) {
          res.v = device_A.in.calibrate.action();
        }
        if (res.v == IDevice.result_t.OK) {
          s = State.Initialized;
          reply_IDevice_result_t = IDevice.result_t.OK;
        }
        else {
          s = State.Uninitialized;
          reply_IDevice_result_t = IDevice.result_t.NOK;
        }
      }
    }
    else if (s == State.Initialized) {
      throw new RuntimeException("illegal");
    }
    else if (s == State.Error) {
      throw new RuntimeException("illegal");
    }
    return reply_IComp_result_t;
  };

  public IComp.result_t client_recover() {
    if (s == State.Uninitialized) {
      throw new RuntimeException("illegal");
    }
    else if (s == State.Initialized) {
      throw new RuntimeException("illegal");
    }
    else if (s == State.Error) {
      {
        V<IDevice.result_t> res = new V <IDevice.result_t>(device_A.in.calibrate.action());
        if (res.v == IDevice.result_t.OK) {
          s = State.Initialized;
          reply_IDevice_result_t = IDevice.result_t.OK;
        }
        else {
          s = State.Error;
          reply_IDevice_result_t = IDevice.result_t.NOK;
        }
      }
    }
    return reply_IComp_result_t;
  };

  public IComp.result_t client_perform_actions() {
    if (s == State.Uninitialized) {
      throw new RuntimeException("illegal");
    }
    else if (s == State.Initialized) {
      {
        V<IDevice.result_t> res = new V <IDevice.result_t>(device_A.in.perform_action1.action());
        if (res.v == IDevice.result_t.OK) {
          res.v = device_A.in.perform_action2.action();
        }
        if (res.v == IDevice.result_t.OK) {
          s = State.Initialized;
          reply_IDevice_result_t = IDevice.result_t.OK;
        }
        else {
          s = State.Error;
          reply_IDevice_result_t = IDevice.result_t.NOK;
        }
      }
    }
    else if (s == State.Error) {
      throw new RuntimeException("illegal");
    }
    return reply_IComp_result_t;
  };

}

File: gaiag/baseline/Comp.c
// Dezyne --- Dezyne command line tools
// Copyright © 2015 Jan Nieuwenhuizen <janneke@gnu.org>
// Copyright © 2015 Rutger van Beusekom <rutger.van.beusekom@verum.com>
//
// This file is part of Dezyne.
//
// Dezyne is free software: you can redistribute it and/or modify it
// under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// Dezyne is distributed in the hope that it will be useful, but
// WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public
// License along with Dezyne.  If not, see <http://www.gnu.org/licenses/>.
//
// Commentary:
//
// Code:

#include "Comp.h"

#include "locator.h"
#include "runtime.h"
#include <assert.h>

typedef enum {
	Comp_State_Uninitialized, Comp_State_Initialized, Comp_State_Error
} Comp_State;




typedef struct {int size;int (*f)(void*);Comp* self;} args_client_initialize;
typedef struct {int size;int (*f)(void*);Comp* self;} args_client_recover;
typedef struct {int size;int (*f)(void*);Comp* self;} args_client_perform_actions;




static void helper_client_initialize(void* args) {
	args_client_initialize *a = args;
	a->f(a->self);
}

static void helper_client_recover(void* args) {
	args_client_recover *a = args;
	a->f(a->self);
}

static void helper_client_perform_actions(void* args) {
	args_client_perform_actions *a = args;
	a->f(a->self);
}







static int client_initialize(void* self_) {
	Comp* self = self_;
	(void)self;
	DZN_LOG("Comp.client_initialize");
	if (self->s == Comp_State_Uninitialized) {
		int res = self->device_A->in.initialize(self->device_A);
		if (res == IDevice_result_t_OK) {
			res = self->device_A->in.calibrate(self->device_A);
		}
		if (res == IDevice_result_t_OK) {
			self->s = Comp_State_Initialized;
			self->reply_IDevice_result_t = IDevice_result_t_OK;
		}
		else {
			self->s = Comp_State_Uninitialized;
			self->reply_IDevice_result_t = IDevice_result_t_NOK;
		}
	}
	else if (self->s == Comp_State_Initialized) {
		assert(false);
	}
	else if (self->s == Comp_State_Error) {
		assert(false);
	}
	return self->reply_IComp_result_t;
}

static int client_recover(void* self_) {
	Comp* self = self_;
	(void)self;
	DZN_LOG("Comp.client_recover");
	if (self->s == Comp_State_Uninitialized) {
		assert(false);
	}
	else if (self->s == Comp_State_Initialized) {
		assert(false);
	}
	else if (self->s == Comp_State_Error) {
		int res = self->device_A->in.calibrate(self->device_A);
		if (res == IDevice_result_t_OK) {
			self->s = Comp_State_Initialized;
			self->reply_IDevice_result_t = IDevice_result_t_OK;
		}
		else {
			self->s = Comp_State_Error;
			self->reply_IDevice_result_t = IDevice_result_t_NOK;
		}
	}
	return self->reply_IComp_result_t;
}

static int client_perform_actions(void* self_) {
	Comp* self = self_;
	(void)self;
	DZN_LOG("Comp.client_perform_actions");
	if (self->s == Comp_State_Uninitialized) {
		assert(false);
	}
	else if (self->s == Comp_State_Initialized) {
		int res = self->device_A->in.perform_action1(self->device_A);
		if (res == IDevice_result_t_OK) {
			res = self->device_A->in.perform_action2(self->device_A);
		}
		if (res == IDevice_result_t_OK) {
			self->s = Comp_State_Initialized;
			self->reply_IDevice_result_t = IDevice_result_t_OK;
		}
		else {
			self->s = Comp_State_Error;
			self->reply_IDevice_result_t = IDevice_result_t_NOK;
		}
	}
	else if (self->s == Comp_State_Error) {
		assert(false);
	}
	return self->reply_IComp_result_t;
}

static int callback_client_initialize(void* self_) {
	Comp* self = ((IComp*)self_)->in.self;
	args_client_initialize a = {sizeof(args_client_initialize), client_initialize, self};
	runtime_event(helper_client_initialize, &a);
	return self->reply_IComp_result_t;
}

static int callback_client_recover(void* self_) {
	Comp* self = ((IComp*)self_)->in.self;
	args_client_recover a = {sizeof(args_client_recover), client_recover, self};
	runtime_event(helper_client_recover, &a);
	return self->reply_IComp_result_t;
}

static int callback_client_perform_actions(void* self_) {
	Comp* self = ((IComp*)self_)->in.self;
	args_client_perform_actions a = {sizeof(args_client_perform_actions), client_perform_actions, self};
	runtime_event(helper_client_perform_actions, &a);
	return self->reply_IComp_result_t;
}


void Comp_init (Comp* self, locator* dezyne_locator) {
	self->rt = dezyne_locator->rt;
	runtime_set(self->rt, self);
	self->s = Comp_State_Uninitialized;
	self->client = &self->client_;
	self->client->in.initialize = callback_client_initialize;
	self->client->in.recover = callback_client_recover;
	self->client->in.perform_actions = callback_client_perform_actions;
	self->client->in.self = self;
	self->device_A = &self->device_A_;
	self->device_A->out.self = self;
}
